sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -507,6 +509,136 @@ case class GpuArrayMax(child: Expression) extends GpuUnaryExpression with Implic
     }
 }
 
+case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with ExpectsInputTypes {
+
+  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.length)(ArrayType)
+
+  @transient override lazy val dataType: DataType = {
+    val fields = children.zip(arrayElementTypes).zipWithIndex.map {
+      case ((expr: NamedExpression, elementType), _) =>
+        StructField(expr.name, elementType, nullable = true)
+      case ((_, elementType), idx) =>
+        StructField(idx.toString, elementType, nullable = true)
+    }
+    ArrayType(StructType(fields), containsNull = false)
+  }
+
+  override def nullable: Boolean = children.exists(_.nullable)
+
+  @transient private lazy val arrayElementTypes =
+    children.map(_.dataType.asInstanceOf[ArrayType].elementType)
+
+  override def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Expression =
+    legacyWithNewChildren(newChildren)
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    val numRows = batch.numRows()
+    val numCols = children.size
+    // A container collects all column vectors in use, to ease the release on exceptions.
+    val workBuffers = mutable.ArrayBuffer[ColumnVector]()
+
+    if (numCols == 0) {
+      return GpuScalar(new GenericArrayData(Array.empty[Any]), dataType)
+    }
+
+    // Prepare input columns
+    closeOnExcept(workBuffers) { _ =>
+      children.foreach { expr =>
+        workBuffers += GpuExpressionsUtils.columnarEvalToColumn(expr, batch).getBase
+      }
+    }
+    val inputs = workBuffers.toArray
+    workBuffers.clear()
+
+    withResource(inputs) { inputs =>
+      // Compute array sizes of input arrays
+      closeOnExcept(workBuffers) { _ =>
+        (0 until numCols).foreach { i =>
+          workBuffers += inputs(i).countElements()
+        }
+      }
+
+      // Pick max array size of each row.
+      // Replace with zero if there exists null among input values.
+      // [1, 3, 5, null, 4]
+      // [2, 4, 3,    8, 0]   => [4, 4, 5, 0, 7]
+      // [4, 2, 3,   10, 7]
+      val maxSize = closeOnExcept(workBuffers) { _ =>
+        val max = withResource(ColumnVector.makeList(workBuffers: _*)) { sizes =>
+          workBuffers.foreach(_.close())
+          workBuffers.clear()
+          sizes.listReduce(SegmentedReductionAggregation.max())
+        }
+        withResource(max) { _ =>
+          withResource(GpuScalar.from(0, IntegerType)) { zero =>
+            max.replaceNulls(zero)
+          }
+        }
+      }
+
+      // Generate offsets from max sizes
+      // [4, 3, 5, 2] => [0, 4, 7, 12, 14]
+      workBuffers += closeOnExcept(workBuffers) { _ =>
```

Comment:
This is so confusing to me. Why close workBuffers if there is an exception if we cleared it out on lines 568 and 569? But we are also adding something new into workBuffers, at the end, but it is just a single value? What is happening here?